yarn2nix: Handle lockfile entries with multiple integrity hashes

Smona opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Some NPM packages provide multiple integrity hashes with different algorithms. These are space separated, like below:

```
tar@^4:
  version "4.4.19"
  resolved "https://registry.yarnpkg.com/tar/-/tar-4.4.19.tgz#2e4d7263df26f2b914dee10c825ab132123742f3"
  integrity "sha1-Lk1yY98m8rkU3uEMglqxMhI3QvM= sha512-a20gEsvHnWe0ygBY8JbxoM4w3SJdhc7ZAuxkLqh+nvNQN2IOt0B5lLgM490X5Hl8FF0dl0tOf2ewFYAlIFgzVA=="
```

Currently, yarn2nix isn't able to handle these lockfile entries, producing an error like `error: hash 'Lk1yY98m8rkU3uEMglqxMhI3QvM= sha512' has wrong length for hash type 'sha1'`. This is because it isn't aware that spaces could separate multiple hashes.

This PR changes yarn2nix to use the [official Standard Subresource Integrity package](https://www.npmjs.com/package/ssri) from npm to parse the integrity line, and pick the best available algorithm. It also replaced many of the local yarn2nix entries with sha512, even ones that don't include it in the lockfile. Not sure how that happened, but it works in practice!

The addition of `ssri` may also reduce signs and symptoms of depression.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

Smona commented:
@jtojnar I hope a direct ping is okay, it looked like you're a good person to review this and i want to make sure this is on someone's radar, since this is a blocker for implementing nix at my work.

I'm still pretty new around here, so I'm not sure what the usual protocol is when a PR doesn't get review requested by ofborg.

Smona committed:
yarn2nix: Handle lockfile entries with multiple integrity hashes

Some NPM packages provide multiple integrity hashes with different
algorithms. These are space separated, like below:

```
tar@^4:
  version "4.4.19"
  resolved "https://registry.yarnpkg.com/tar/-/tar-4.4.19.tgz#2e4d7263df26f2b914dee10c825ab132123742f3"
  integrity "sha1-Lk1yY98m8rkU3uEMglqxMhI3QvM= sha512-a20gEsvHnWe0ygBY8JbxoM4w3SJdhc7ZAuxkLqh+nvNQN2IOt0B5lLgM490X5Hl8FF0dl0tOf2ewFYAlIFgzVA=="
```

Currently, yarn2nix isn't able to handle these lockfile entries,
producing an error like `error: hash 'Lk1yY98m8rkU3uEMglqxMhI3QvM=
sha512' has wrong length for hash type 'sha1'`. This is because it isn't
aware that spaces could separate multiple hashes.

This commit uses the official Standard Subresource Integrity package
from npm to parse the integrity line, and pick the best available
algorithm. It also replaced many of the local yarn2nix entries with
sha512, even ones that don't include it in the lockfile. Not sure how
that happened, but it works in practice!

The addition of `ssri` may also reduce signs and symptoms of depression.

Smona committed:
yarn2nix: format code

Smona commented:
Thanks! Commit messages are fixed; I've gotten too used to the squash merge with copied PR description workflow.

Thanks also for the nudge towards correctness. Rather than naively picking the last digest, this now uses the [official Standard Subresource Integrity package](https://www.npmjs.com/package/ssri#integrity-to-string) to parse the integrity and choose the best algorithm. I re-ran my test case (and dogfooded the local yarn2nix generation) and it still works now.

I also ran the formatting script for my changes, but it caught a bunch of other formatting issues. I figured it would be best to include that.

jtojnar commented on pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/generateNix.js:
```diff
@@ -103,7 +104,14 @@ function fetchLockedDep(builtinFetchGit) {
+    let algo = "sha1";
+    let hash = sha1OrRev;
+    if (integrity) {
+      const integrities = ssri.parse(integrity);
```
This looks good but apparently, it might choose unsupported hash type. Nix [supports](https://nixos.org/manual/nix/stable/language/advanced-attributes.html#adv-attr-outputHashAlgo) `sha1`, `sha256` and `sha512`, while [`ssri.parse`](https://github.com/npm/ssri#parse) can return ¿anything? (or `sha256`, `sha384`, and `sha512`) in strict mode).

Smona replied:
Thanks, I wasn't aware of that constraint! I almost threw out the `ssri` dependency to use a simple filter & sort method after learning that, but decided to leave it. This way we can handle parsing less-common stuff like integrity options, and it will be more obvious how to pick the best algorithm if nix supports new ones in the fewer.

RaitoBezarius replied:
It does not seem this is fixed, is it?

RaitoBezarius replied:
Nevermind, was reading commit per commit

jtojnar commented on pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/generateNix.js:
```diff
@@ -103,7 +104,14 @@ function fetchLockedDep(builtinFetchGit) {
+    if (integrity) {
+      const integrities = ssri.parse(integrity);
+      algo = integrities.pickAlgorithm();
+      hash = integrities[algo][0].hexDigest();
```
Nix supports `base64` so it might be nicer to use that.

jtojnar commented on pkgs/development/tools/yarn2nix-moretea/yarn2nix/bin/yarn2nix.js:
```diff
@@ -1,14 +1,14 @@
 #!/usr/bin/env node
```
Would be nice to mention how it was formatted in the commit message.

I believe the choice of single quotes and omission of return values was intentional but if we can rely on auto-formatter, that will allow us to write code without having to think about low level things like syntax so :+1: from me.

Smona replied:
I just used the `yarn format` script in the package.json. I completely agree, I don't really care what the code style is, as long as it's encoded into a format command and enforced so the code looks good and you don't have to think about it.

jtojnar commented on pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/generateNix.js:
```diff
@@ -34,18 +34,24 @@ const { execFileSync } = require('child_process')
-      let${builtinFetchGit ? `
+      let${
+        builtinFetchGit
+          ? `
         repo = builtins.fetchGit ({
```
Maybe we should extract this into a variable rather than nest the format-strings so that the indentation does not jump weirdly like this.

Smona committed:
yarn2nix: limit ssri parsing to supported algorithms

Nix supports md5, sha1, sha256, and sha512 hashes. The first two are
deprecated, but many npm packages provide only a sha1, so we have to
allow that one. This commit continues to leverage `ssri` for reliable
integrity string parsing and algorithm selection, but limits it to only
supported hash algorithms.

It also switches the hash representation in the generated nix file to
base64, since that format is supported by nix (suggested by jtojnar).
This saves a little disk space and computation.

Finally, it adds a .prettierignore file, since prettier called with the
arguments in the format command currently tries to format node_modules.

Smona commented:
Currently running `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`

Rebuilt packages:

apache-airflow CharacterCompressor CompBus constant-detune-chorus discourse discourse element-desktop element-desktop element-web emscripten emscripten-json-c emscripten-libxml2 emscripten-xmlmirror-unstable emscripten-zlib faust faust2alqt faust2alsa faust2csound faust2firefox faust2jack faust2jackrust faust2jaqt faust2ladspa faust2lv2 faustCompressors faustlive faustPhhysicalModeling faustPhysicalModeling fixup_yarn_lock gitlab gitlab-ee gotify-server guitarix HedgeDoc HedgeDoc heroic heroic kapitonov-plugins-pack LazyLimiter lemmy-ui listmonk mastodon matrix-alertmanager matrix-appservice-discord matrix-appservice-slack MBdistortion mephisto meshcentral micropad mirakurun mooSpace-unstable oni2 outline peertube pgadmin pgadmin plausible pluginUtils pomerium powerdns-admin puppeteer-cli python3.10-apache-airflow python3.10-tandoor-recipes python3.9-apache-airflow ratman renderer RhythmDelay rmfakecloud RStudio RStudio RStudio RStudio schildichat-desktop schildichat-desktop schildichat-web Sharedown shelfMultiBand synapse-admin Tambura uivonim-unstable vieb VoiceOfFaust woodpecker-server yarn2nix zammad

Smona commented:
Okay, the oni2 build is failing in the exact same way as the master branch now 😅

jtojnar commented on pkgs/applications/editors/oni2/default.nix:
```diff
@@ -11,6 +11,6 @@ in mkOni2 rec {
   rev = "v${version}";
   sha256 = "NlN0Ntdwtx5XLjd1ltUzv/bjmJQR5eyRqtmicppP6YU=";
-  fetchDepsSha256 = "k7G6jPJfxCCSuSucPfiXljCVJhmjl/BxWMCEjv2tfhA=";
+  fetchDepsSha256 = "jtPRgAFyoy2ILPQyHVVIzX2DZj30K6xnJei1P4SMtBc=";
```
It seems to fail for me:

```
error: hash mismatch in fixed-output derivation '/nix/store/zwfppn42k5i6f2cvbv23ydlfg0wz86z0-oni2-fetch-deps.drv':
         specified: sha256-jtPRgAFyoy2ILPQyHVVIzX2DZj30K6xnJei1P4SMtBc=
            got:    sha256-k7G6jPJfxCCSuSucPfiXljCVJhmjl/BxWMCEjv2tfhA=
```

Smona replied:
not quite sure what to make of this... ftr the new hash fails for me too in this branch but passes in nixpkgs-review.

<details>
<summary>Output:</summary>
<pre>
❯ diffoscope /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps
--- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps
+++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps
│   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy
├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy
│ ├── file list
│ │ @@ -1,3 +1,3 @@
│ │  3
│ │ +3__________________________
│ │ -3_________________________________________________________________________
│ │  source
│ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source
│ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source
│ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i
│ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i
│ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/lodash__4.17.21__82c45c9d
│ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/lodash__4.17.21__82c45c9d
│ │ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/lodash__4.17.21__82c45c9d/fp
│ │ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/lodash__4.17.21__82c45c9d/fp
│ │ │ │ │ ├── stat {}
│ │ │ │ │ │ @@ -1,7 +1,7 @@
│ │ │ │ │ │  
│ │ │ │ │ │ +  Size: 16384      Blocks: 32         IO Block: 4096   directory
│ │ │ │ │ │ -  Size: 20480      Blocks: 40         IO Block: 4096   directory
│ │ │ │ │ │  Device: 8,32       Access: (0555/dr-xr-xr-x)  Uid: ( 1000/   smona)   Gid: ( 1000/   smona)
│ │ │ │ │ │  
│ │ │ │ │ │  Modify: 1970-01-01 00:00:01.000000000 +0000
│ │ │ │ │ ├── stat {}
│ │ │ │ │ │ @@ -1,7 +1,7 @@
│ │ │ │ │ │  
│ │ │ │ │ │ -  Size: 20480      Blocks: 40         IO Block: 4096   directory
│ │ │ │ │ │ +  Size: 16384      Blocks: 32         IO Block: 4096   directory
│ │ │ │ │ │  Device: 8,32       Access: (0555/dr-xr-xr-x)  Uid: ( 1000/   smona)   Gid: ( 1000/   smona)
│ │ │ │ │ │  
│ │ │ │ │ │  Modify: 1970-01-01 00:00:01.000000000 +0000
│ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/opam__s__omd__337d3e7d
│ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/opam__s__omd__337d3e7d
│ │ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/opam__s__omd__337d3e7d/tests
│ │ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/opam__s__omd__337d3e7d/tests
│ │ │ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/opam__s__omd__337d3e7d/tests/commonmark
│ │ │ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/opam__s__omd__337d3e7d/tests/commonmark
│ │ │ │ │ │ ├── stat {}
│ │ │ │ │ │ │ @@ -1,7 +1,7 @@
│ │ │ │ │ │ │  
│ │ │ │ │ │ │ -  Size: 45056            Blocks: 88         IO Block: 4096   directory
│ │ │ │ │ │ │ +  Size: 36864            Blocks: 72         IO Block: 4096   directory
│ │ │ │ │ │ │  Device: 8,32     Access: (0555/dr-xr-xr-x)  Uid: ( 1000/   smona)   Gid: ( 1000/   smona)
│ │ │ │ │ │ │  
│ │ │ │ │ │ │  Modify: 1970-01-01 00:00:01.000000000 +0000
│ │ │ │ │ │ ├── stat {}
│ │ │ │ │ │ │ @@ -1,7 +1,7 @@
│ │ │ │ │ │ │  
│ │ │ │ │ │ │ +  Size: 36864            Blocks: 72         IO Block: 4096   directory
│ │ │ │ │ │ │ -  Size: 45056            Blocks: 88         IO Block: 4096   directory
│ │ │ │ │ │ │  Device: 8,32     Access: (0555/dr-xr-xr-x)  Uid: ( 1000/   smona)   Gid: ( 1000/   smona)
│ │ │ │ │ │ │  
│ │ │ │ │ │ │  Modify: 1970-01-01 00:00:01.000000000 +0000
│ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5
│ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5
│ │ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfbuzz-2.6.
8
│ │ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfbuzz-2.6.
8
│ │ │ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfbuzz-2.
6.8/test
│ │ │ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfbuzz-2.
6.8/test
│ │ │ │ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfbuzz-
2.6.8/test/subset
│ │ │ │ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfbuzz-
2.6.8/test/subset
│ │ │ │ │ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfbuz
z-2.6.8/test/subset/data
│ │ │ │ │ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfbuz
z-2.6.8/test/subset/data
│ │ │ │ │ │ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfb
uzz-2.6.8/test/subset/data/expected
│ │ │ │ │ │ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/harfb
uzz-2.6.8/test/subset/data/expected
│ │ │ │ │ │ │ │ │ │   --- /nix/store/fkyri627k46clh3mqy9vq9qxbnprzsdx-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/har
fbuzz-2.6.8/test/subset/data/expected/cbdt
│ │ │ │ │ │ │ │ │ ├── +++ /nix/store/dvzk5xy24jvqpkmaibw9zfp89yh7xqlc-oni2-fetch-deps/esy/source/i/revery__s__esy_harfbuzz__2.6.8002__3348a4f5/har
fbuzz-2.6.8/test/subset/data/expected/cbdt
│ │ │ │ │ │ │ │ │ │ ├── stat {}
│ │ │ │ │ │ │ │ │ │ │ @@ -1,7 +1,7 @@
│ │ │ │ │ │ │ │ │ │ │  
│ │ │ │ │ │ │ │ │ │ │ -  Size: 12288            Blocks: 24         IO Block: 4096   directory
│ │ │ │ │ │ │ │ │ │ │ +  Size: 16384            Blocks: 32         IO Block: 4096   directory
│ │ │ │ │ │ │ │ │ │ │  Device: 8,32     Access: (0555/dr-xr-xr-x)  Uid: ( 1000/   smona)   Gid: ( 1000/   smona)
│ │ │ │ │ │ │ │ │ │ │  
│ │ │ │ │ │ │ │ │ │ │  Modify: 1970-01-01 00:00:01.000000000 +0000
│ │ │ │ │ │ │ │ │ │ ├── stat {}
│ │ │ │ │ │ │ │ │ │ │ @@ -1,7 +1,7 @@
│ │ │ │ │ │ │ │ │ │ │  
│ │ │ │ │ │ │ │ │ │ │ +  Size: 16384            Blocks: 32         IO Block: 4096   directory
│ │ │ │ │ │ │ │ │ │ │ -  Size: 12288            Blocks: 24         IO Block: 4096   directory
│ │ │ │ │ │ │ │ │ │ │  Device: 8,32     Access: (0555/dr-xr-xr-x)  Uid: ( 1000/   smona)   Gid: ( 1000/   smona)
│ │ │ │ │ │ │ │ │ │ │  
│ │ │ │ │ │ │ │ │ │ │  Modify: 1970-01-01 00:00:01.000000000 +0000
</pre>
</details>

Smona replied:
Update: I created a copy of this branch and rebased it to master as of this comment. Then the build succeeds with the new hash. I guess nixpkgs-review works!

(i also accidentally synced this branch on my fork rather than master first, so force-pushed to undo that merge commit)

jtojnar replied:
It think that hash mismatch might be unrelated to this change and the issue might be that the package is non-deterministic.

Smona replied:
i think you're right. so, does the PR look good to go now?

jtojnar replied:
I would drop the oni2 change.

Smona replied:
done :heavy_check_mark:

Smona commented:
@jtojnar since tagging you here I've seen you all over the place so I know you must be busy, is there anything I can do to help move this forward?

jtojnar commented on pkgs/development/tools/yarn2nix-moretea/yarn2nix/.prettierignore:
```diff
@@ -0,0 +1 @@
+node_modules
```
According to https://prettier.io/docs/en/ignore.html prettier should ignore `node_modules` by default?

Smona replied:
It's supposed to, but in practice it was running on node_modules without the `.prettierignore`. Perhaps the prettier version is pinned before https://github.com/prettier/prettier/pull/2185 went in, or perhaps it's something in my environment, but it seemed like a safe change.

jtojnar reviewed (approved):
LGTM but I was only able to look at the code briefly.

RaitoBezarius commented:
Can I do something to help you getting this merged?

Smona commented:
@RaitoBezarius thanks for jumping in! I don't have write access so I'm just waiting for a maintainer to merge this. The code is ready to go.

RaitoBezarius commented on pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/generateNix.js:
```diff
@@ -46,37 +46,55 @@ function prefetchgit(url, rev) {
+  }
+
+  algo = integrities.pickAlgorithm();
+  hash = integrities[algo][0].digest;
```
Weren't we supposed to use `hexDigest()` ?

Smona replied:
previously we were using hex digests because that's what is in the integrity line on the lock file, and the only processing we would do on that is splitting the digest and Algo.

however, since SSRI already parses the digests into base64, [jtojnar suggested](https://github.com/NixOS/nixpkgs/pull/197185#discussion_r1006475038) suggested we just use that formay, since nix supports it. base64 is much more compact than base16, so it will save a little disk space for large yarn projects

RaitoBezarius reviewed (approved):
I'm not an expert with yarn2nix, so I will be running a full-blown nixpkgs-review on the top of the review. Seems good modulo some minor questions.

RaitoBezarius commented:
Result of `nixpkgs-review pr 197185` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package marked as broken and skipped:</summary>
  <ul>
    <li>micropad</li>
  </ul>
</details>
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>faustStk</li>
    <li>magnetophonDSP.ConstantDetuneChorus</li>
    <li>mirakurun</li>
    <li>oni2</li>
  </ul>
</details>
<details>
  <summary>73 packages built:</summary>
  <ul>
    <li>apache-airflow (python310Packages.apache-airflow)</li>
    <li>discourse</li>
    <li>discourseAllPlugins</li>
    <li>element-desktop</li>
    <li>element-desktop-wayland</li>
    <li>element-web</li>
    <li>emscripten</li>
    <li>faust (faust2)</li>
    <li>faust2alqt</li>
    <li>faust2alsa</li>
    <li>faust2csound</li>
    <li>faust2firefox</li>
    <li>faust2jack</li>
    <li>faust2jackrust</li>
    <li>faust2jaqt</li>
    <li>faust2ladspa</li>
    <li>faust2lv2</li>
    <li>faustPhysicalModeling</li>
    <li>faustlive</li>
    <li>fixup_yarn_lock</li>
    <li>gitlab</li>
    <li>gitlab-ee</li>
    <li>gotify-server</li>
    <li>grafana-image-renderer</li>
    <li>guitarix</li>
    <li>hedgedoc</li>
    <li>heroic</li>
    <li>heroic-unwrapped</li>
    <li>kapitonov-plugins-pack</li>
    <li>lemmy-ui</li>
    <li>listmonk</li>
    <li>magnetophonDSP.CharacterCompressor</li>
    <li>magnetophonDSP.CompBus</li>
    <li>magnetophonDSP.LazyLimiter</li>
    <li>magnetophonDSP.MBdistortion</li>
    <li>magnetophonDSP.RhythmDelay</li>
    <li>magnetophonDSP.VoiceOfFaust</li>
    <li>magnetophonDSP.faustCompressors</li>
    <li>magnetophonDSP.pluginUtils</li>
    <li>magnetophonDSP.shelfMultiBand</li>
    <li>mastodon</li>
    <li>matrix-alertmanager</li>
    <li>matrix-appservice-discord</li>
    <li>matrix-appservice-slack</li>
    <li>meshcentral</li>
    <li>mooSpace</li>
    <li>open-music-kontrollers.mephisto</li>
    <li>outline</li>
    <li>peertube</li>
    <li>pgadmin4</li>
    <li>plausible</li>
    <li>pomerium</li>
    <li>powerdns-admin</li>
    <li>puppeteer-cli</li>
    <li>python39Packages.apache-airflow</li>
    <li>ratman</li>
    <li>rmfakecloud</li>
    <li>rstudio</li>
    <li>rstudio-server</li>
    <li>rstudioServerWrapper</li>
    <li>rstudioWrapper</li>
    <li>schildichat-desktop</li>
    <li>schildichat-desktop-wayland</li>
    <li>schildichat-web</li>
    <li>sharedown</li>
    <li>synapse-admin</li>
    <li>tambura</li>
    <li>tandoor-recipes</li>
    <li>uivonim</li>
    <li>vieb</li>
    <li>woodpecker-server</li>
    <li>yarn2nix</li>
    <li>zammad</li>
  </ul>
</details>

RaitoBezarius commented:
> 

@Smona Could you check if the 4 packages which failed to build are broken on master too?

Smona commented:
@RaitoBezarius sure, the magnetophone, oni2, and mirakurun failures are already confirmed to be unrelated above but i can take a look at faustStk

RaitoBezarius commented:
> @RaitoBezarius sure, the magnetophone, oni2, and mirakurun failures are already confirmed to be unrelated above but i can take a look at faustStk

Awesome, thank you!

Smona commented:
@RaitoBezarius Indeed, `nix-build -A faustStk` fails on master (9f0ebc3f631) with the following output:

```
bass.cpp:13673:10: fatal error: bass.h: No such file or directory
13673 | #include <bass.h>
      |          ^~~~~~~~
compilation terminated.
make: *** [Makefile:223: bass.o] Error 1
error: builder for '/nix/store/367six8x2dcgva3c97w32x240bimly44-faustPhhysicalModeling-2.20.2.drv' failed with exit code 2;
       last 10 log lines:
       >          end prepare2 (duration : 6.91414e-05)
       >
       >         end compileMultiSignal (duration : 0.069227)
       >
       > end generateCode (duration : 0.10184)
       > bass.cpp:13673:10: fatal error: bass.h: No such file or directory
       > 13673 | #include <bass.h>
       >       |          ^~~~~~~~
       > compilation terminated.
       > make: *** [Makefile:223: bass.o] Error 1
       For full logs, run 'nix log /nix/store/367six8x2dcgva3c97w32x240bimly44-faustPhhysicalModeling-2.20.2.drv'.

```
